src/Kernel/Magic/DebugMagic.cs

```diff
@@ -0,0 +1,224 @@
+        }
+
+        private const string ParameterNameOperationName = "__operationName__";
+        private ConcurrentDictionary<Guid, ManualResetEvent> sessionAdvanceEvents
+            = new ConcurrentDictionary<Guid, ManualResetEvent>();
+        private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+        private IShellServer shellServer;
+        private ILogger<DebugMagic> logger;
+
+        /// <summary>
+        ///      The symbol resolver used by this magic command to find
+        ///      operations or functions to be simulated.
+        /// </summary>
+        public ISymbolResolver SymbolResolver { get; }
+
+        /// <summary>
+        ///     The configuration source used by this magic command to control
+        ///     simulation options (e.g.: dump formatting options).
+        /// </summary>
+        public IConfigurationSource ConfigurationSource { get; }
+
+        /// <inheritdoc />
+        public override ExecutionResult Run(string input, IChannel channel) =>
+            RunAsync(input, channel).Result;
+
+        private async Task HandleStartMessage(Message message) => await Task.Run(() =>
+        {
+            var content = (message.Content as UnknownContent);
+            var session = content?.Data?["debug_session"];
+            if (session == null)
+            {
+                // TODO: Communicate error back here!
+            }
+            else
+            {
+                shellServer.SendShellMessage(
+                    new Message
+                    {
+                        Header = new MessageHeader
+                        {
+                            MessageType = "iqsharp_debug_reply"
+                        },
+                        Content = new UnknownContent
+                        {
+                            Data = new Dictionary<string, object>
+                            {
+                                ["debug_session"] = session
+                            }
+                        }
+                    }
+                    .AsReplyTo(message)
+                );
+            }
+        });
+
+        private async Task HandleAdvanceMessage(Message message) => await Task.Run(() =>
+        {
+            logger.LogDebug("Got debug advance message:", message);
+            var content = (message.Content as UnknownContent);
+            var session = content?.Data?["debug_session"];
+            if (session == null)
+            {
+                logger.LogWarning("Got debug advance message, but debug_session was null.", message);
+            }
+            else
+            {
+                var sessionGuid = Guid.Parse(session.ToString());
+                ManualResetEvent? @event = null;
+                lock (sessionAdvanceEvents)
+                {
+                    @event = sessionAdvanceEvents[sessionGuid];
+                }
+                @event.Set();
```

Comment:
Would this throw an error if `@event` is `null`?

Reply:
Same for L156/166.

Reply:
To be completely honest, I'm not sure...